# Сборник 2 из 19-ти задач. Соколова Александра  
---
## Из раздела "Я люблю математику"
### Теория вероятностей и мат. статистика

**Задача 1:**  
Амеба имеет шанс 25%, 25% и 50% произвести 0, 1 или 2 потомства соответственно. Каждый из потомков амебы также имеет одинаковые вероятности.  
Какова вероятность того, что род амебы вымрет?  

**Решение для задачи 1:** 
Допустим искомая вероятность вымирания рода амебы равна **p**.  
Событие вымирания рода случается если случится одно из трех зависимых событий перечисленных ниже:  
1-ое событие (А) - с вероятностью 25% сразу не будет потомства => P(A) = 1/4 - вероятность события А  
2-ое событие (В)- с вероятностью 25% родиться одна амеба и у нее с вероятностью **p** вымрет род => P(B)=1/4***p**  
3-е событие (С) - с вороятностью 50% родится две амебы и у них вымрут рода одновременно => P(C) = 1/2***p**^2 - поттому что потомки амебы имеют одинаковую с родителем вероятность вымирания рода  
P(события вымирания рода) = P(A) + P(B) + P(C), так как у одной амебы может родиться одновременно максимально два потомка  
P(события вымирания рода) = **p**  

**p** = 1/4 + 1/4***p** + 1/2* **p**^2  
получаем квадратное уравнение: 1/2* **p**^2- 3/4***p** + 1/4  = 0  
2 **p**^2 - 3 **p** + 1 = 0  
D(Determinant) = 3*3 - 4*2*1 = 1  
**p** = (3 ∓ √D)/(2*2) = (3 ∓ 1)/4 = 1 и 0.5  

Ответ **p** = 1 отбрасываем так как это вырожденный случай.  

**Ответ для задачи 1: p= 0.5**   

**Задача 2:**  
В любой 15-минутный интервал с вероятностью 20% вы увидите хотя бы одну падающую звезду.  
Какова вероятность того, что вы увидите хотя бы одну падающую звезду в течение часа?  

**Решение для задачи 2:**  
Будем использовать формулу Бернули:  
$P_n(m) = C_n^m p^m q^{(n-m)}$  
$C_n^k = \frac{n!}{k! (n-k)!}$  

Вероятность увидеть звезду в течение 15 минут = 0.2 = p (можем прдположить что это один опыт который проходит в течении 15 минут и вероятность успеха опыта равна p)  
Вероятность не увидеть звезду в течение 15 минут = q = 1 - p = 1 - 0.2 = 0.8  

Проведем n=4 опытов (4*15 минут = 60 минут = 1 час), в каждом из которых может произойти определенное событие («успех») с вероятностью p (или не произойти — «неудача» — с вероятностью q=1-p. Задача — найти вероятность получения ровно m = 0 успехов в этих n=4 опытах.  

$P_4(0) = C_4^0 p^0 q^{(4-0)} = (0.8)⁴ $  

Вероятность того, что вы НЕ увидите НИ одну падающую звезду в течение часа = Вероятность не увидеть ни одной звезды за 15 минут 4 раза подряд = (0.8)⁴ = 0.4096 так как падение звезд в каждый 15-ти минутный промежуток времени независимо друг от друга  

Вероятность того, что вы увидите хотя бы одну падающую звезду в течение часа = 1 - Вероятность того, что вы НЕ увидите НИ одну падающую звезду в течение часа = 1 - 0.4096 = 0,5904  

**Ответ для задачи 2: 0,5904**   

**Дополнительное решение для задачи 2:**  
В сети я увидел еще одно решение этой задачи с помощью формулы Пуассона, но на мой взгляд её использование не собосновано, потому что одно из требований применения формулы достаточно большое n. В некоторых источниках указывается что n должно быть больше 30, а n*p <10. Что в рамках нашей задчи не выполняется  

In [78]:
# решение задачи 2 с помощью формулы Пуассона
import scipy.stats as ss
1-ss.poisson.cdf(0,0.8)

0.5506710358827783

**Задача 3:**  
Как вы можете сгенерировать случайное число от 1 до 7, используя только кубик?  

**Решение для задачи 3:**  
Для начала договоримся, что кубик у нас идеальный. При бросании кубика случайная величина — число точек на грани —   
принимает одно из 6-и возможных значений: {1,2,3,4,5,6}. Вероятность выпадения одной точки из шести равна 1/6,  
одинакова для каждой точки, поэтому случайная величина имеет дискретное равномерное распределение со следующими параметрами:  
мат. ожидание = 3.5,  
дисперсия ~= 2.916666,  
стандартное отклонение ~= 1.7078,  
коэффициент асимметрии = 0,  
коэффициент эксцесса ~= -1.2685  

Нам необходимо создать способ транформации вышеописанного распределения в распределение с возможными значениями: {1,2,3,4,5,6,7}.  
Вероятность выпадения одного значения из семи равна 1/7 и одинакова для каждого значения и случайная величина имеет дискретное   
равномерное распределение со следующими параметрами:  
мат. ожидание = 4,  
дисперсия = 4,  
стандартное отклонение = 4,  
коэффициент асимметрии = 0,  
коэффициент эксцесса ~= -1.25  

**Ответ для задачи 3**   
Способ который решает вышеописанную задачу:  
Бросаем кубик 3 раза. Каждый раз интерпретируем выпадаемое на кубике значение, как четное или нечетное, то есть переводим в биты.  
Выпадает четное число это 0 (ноль), выпадает нечетное это - 1 (один).  
Получаем три бита (бит - это один разряд двоичного кода (двоичная цифра)), которые начинаются от '000' и заканчиваются '111'.  
Двоичное число '000' соответствует десятичному числу 0(ноль) ,а двоичное число '111' соответствует десятичному числу 7.  
Получаем пространство элементарных исходов, состоящим из 8-ми (восьми) элементов: {0, 1, 2, 3, 4, 5, 6, 7}.   
Полученные элементарные исходы равновозможны так как исходное распределение дискретное равномерное.  
Но нам нужно не восем, а семь исходов. Поэтому мы договариваемся, что если мы получаем исход со значением 0,   
то мы перебрасываем кубики. Тут важно отметить, что процесс с какой-то вероятностью может не завершиться никогда,   
потому что могут непрерывно выпадать четные числа и мы будем получать 000 и будем вынуждены перебрасывать снова и снова.   
Можно посчитать, что эта вероятность равна 1/2 в степени n, где n кол-во подбрасываний. Что означает что при n больше 20 она будет ничтожно мала (меньше 10**-7)  
Таким образом мы получим пространство исходов {1, 2, 3, 4, 5, 6, 7}  



In [84]:
# давайте проверим этот способ
# сначала смоделируем распределение соответсвующее экспериментам по бросанию кубика
from scipy.stats import randint
low, high = 1, 7
mean, var, skew, kurt = randint.stats(low, high, moments='mvsk')  # проверяем параметры дистректного равномерного распределения
                                                                  # которое соответствует бросанию игрального кубика
print('Мат.ожидание: ', mean)
print('Дисперсия: ', var)
print('Стандартное отклонение: ', var**1/2)
print('Коэффициент асимметрии: ', skew)
print('Коэффициент эксцесса: ', kurt)

Мат.ожидание:  3.5
Дисперсия:  2.9166666666666665
Стандартное отклонение:  1.4583333333333333
Коэффициент асимметрии:  0.0
Коэффициент эксцесса:  -1.2685714285714285


In [108]:
# сгенерируем массив из десяти миллионов (10 000 000) бросаний кубика
r = randint.rvs(low, high, size=10000000)
r

array([1, 6, 1, ..., 4, 3, 1])

In [109]:
# проверим уникальные значения, среднее и стандартное отклонение этого массива
import pandas as pd
df_r = pd.DataFrame(data= r)
print('Уникальные значения:', sorted(df_r[0].unique()))
df_r.describe()


Уникальные значения: [1, 2, 3, 4, 5, 6]


0
count  1.000000e+07
mean   3.500421e+00
std    1.708246e+00
min    1.000000e+00
25%    2.000000e+00
50%    4.000000e+00
75%    5.000000e+00
max    6.000000e+00

In [113]:
# посмотрим на вероятности появления каждого значения
df_r[0].value_counts(normalize=True)

6    0.166919
4    0.166746
2    0.166736
1    0.166683
5    0.166511
3    0.166405
Name: 0, dtype: float64

In [114]:
# проверим выполнение классического определения вероятности для полученного распределения
df_r[0].value_counts(normalize=True).sum()

1.0

In [118]:
# теперь создадим функцию которая будет реализовывать наш способ и транформировать одно распределение в другое
import numpy as np

def conver_uniform_1_6_to_uniform_1_7 (d_r: np.array):
    new_r = np.zeros(len(d_r))
    count, count_zero = 0, 0
    bite_ = ''
    for i in range(len(d_r)):
        bite_ += str(int(d_r[i]<4))
        if len(bite_) == 3:
            a = int(bite_,2) 
            if a > 0:
                new_r[count] = int(bite_,2)
                count += 1
            else:
                count_zero += 1
            bite_ = ''
    return new_r, count, count_zero

In [120]:
new_r, count, count_zero = conver_uniform_1_6_to_uniform_1_7(r)

In [121]:
new_r = new_r[:count]
len(new_r)

2916199

**Замечание:**  
из 10-ти миллионов подбрасываний мы получили 2916199 значений нового распределения. Получается, что если бы мы не отбрасывали нулевые значения мы должны были получить 10000000/3 = 3333333 значения. А получили 2916199. Значит нули были в 3333333-2916199 = 417134 случаях. Это почти 1/8 (~=417134/3333333) от 3333333 значений, которые мы должны были получить. Что соответсвует доли одного значения, которое мы отбрасываем.  
Важно понимать, что отбрасывание варианта '000' не изменяет структуру равномерного дискретного распределения, а только усложняет алгоритм и требует проверки его трудоемкости, чтобы не сталкиваться с высокими оценками по времени и сложности.  

In [122]:
count_zero

417134

In [127]:
# теперь проверим параметры полученного распределения и сравним с заданным в условии задачи
df_r_new = pd.DataFrame(data= new_r)
print('Уникальные значения:', sorted(df_r_new[0].unique()))
df_r_new.describe()

Уникальные значения: [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0]


0
count  2.916199e+06
mean   3.999537e+00
std    1.999918e+00
min    1.000000e+00
25%    2.000000e+00
50%    4.000000e+00
75%    6.000000e+00
max    7.000000e+00

**Задача 4:**  
Как с помощью не симметричной монеты можно добиться справедливого подбрасывания монеты (как при использовании идеальной монеты)?  
Не симметричная монета - монета у которой вероятность выпадения орла **p** НЕ равна вероятности выпадения решки **q**.  
В идеальной (симметричной) монете вероятность выпадения орла **p** равна вероятности выпадения решки **q** (**p** = **q** = 1/2)  

Наша задача придумать способ при котором события, которые мы будем получать будут составлять пространство двух элементарных исходов,  
равновероятных (вероятность одного исхода **p** будет равна вероятности второго исхода **q** и равны 1/2, исходя из классического   
определения вероятности **p** + **q** = 1)

**Ответ для задачи 4**   
Способ позволяющий получить два равновероятных исхода:  
1. подбросим монетку два раза.  
2. если получили одинаковые последовательности ("орел" - "орел" или "решка" - "решка"), то перебрасываем еще раз два раза как в пункте 1.  
3. если получили комбинацию "орел" - "решка", то считаем это выпадением "орел2" и бросаем снова (переходим к пункту 1). Если у нас уже был вариант с "орел2", то перебрасываем снова, до тех пор пока не выпадет вариант "решка2".  
4. если получили комбинацию "решка" - "орел", то считаем это выпадением "решка2" и бросаем снова (переходим к пункту 1). Если у нас уже был вариант с "решка2", то перебрасываем снова, до тех пор пока не выпадет вариант "орел2".  
5. получив один вариант "орел2" и "решка2" мы получим два равновероятных события.  

При этом важно заметить, что если перед вами стоит практическая задача - такая например как определить какой команде будет отдано право разводить мяч в начале футболного матча. То достаточно будет в начале жеребьевки определить какой команде соответсвует комбинация "орел2", а какой "решка2". И далее достаточно получить только одну из этих комбинаций. Нет необходимости получать оба варианта, достаточно будет получить выпадение одного из этих вариантов комбинаций, чтобы определить победителя.  

**Пояснения к ответу для задачи 4**   
Давайте сформулируем задачу в математичсекой области: определим вероятность выпадения орла как **p**. Тогда вероятность выпадения 
решки **q** = 1 - **p**  
Теперь простраство событий двукраного подбрасывания монетки состоит их следующих исходов:
1. орел - орел . Вероятность этого события **p*****p**, так как последовательные подбрасывания монетки это независимые события с вероятностью **p**. Вероятность получить на первом и на втором подбрасывании орел равна **p** умноженное на **p**.
2. орел - решка . Вероятность этого события **p***(1 - **p**)
3. решка - орел . Вероятность этого события (1 - **p**)***p**
4. решка - решка . Вероятность этого события (1 - **p**)*(1 - **p**)

Можно заметить, что вероятность выпадения орла, а затем решки такая же, как вероятность выпадения орла, а затем орла, поскольку мы бросаем одну и туже монету и эти два броска независимы. При исключении исходов «два орла» и «две решки» путем повторения процедуры (двойное подбрасывание) у нас остаются только два оставшихся исхода с равной вероятностью **p***(1 - **p**) = (1 - **p**)***p**.  
Эта процедура работает только в том случае, если броски правильно реализованы парами. То есть если вы вдруг получили варианты "орел" - "орел" и делая следующее подбрасывание вы получили "решку" и пытаетесь назвать это комбинацией 2 - "орел" - "решка" ("орел2"), то вы нарушаете правила эксперимента. Часть пары не должна повторно использоваться в другой паре, честность (точность) эксперимента в таком  случае может быть нарушена.  

Исходы "орел"-"решка"  и "решка"-"орел" равновероятны, без какой-либо предвзятости мы можем назвать, что если происходит событие "орел"-"решка" - это означает "орел2", "решка"-"орел" означает "решка2", но если происходит "орел"-"орел" или "решка"-"решка" мы повторяем эксперимент с двойным подбрасыванием.

Теперь давайте попробуем найти ожидаемое количество подбрасываний монеты, необходимое для того, чтобы определить орел2 или решку2?  
Пусть кол-во ожидаемых бросков монеты равно **n**.  

Вероятность того, что мы получим результат сразу после двух подбрасываний (исходы 2 или 3), равна 2**p**(1 - **p**)  
Общее количество подбрасываний монеты будет равно 2  

Вероятность того, что мы не получим никакого результата после двух подбрасываний, равна (1 - 2**p**(1-**p**))
Общее количество подбрасываний монеты будет 2 + **n** (мы потратили впустую 2 подбрасывания монеты и все равно ожидаем **n** подбрасываний)  

**n** = 2 * 2**p**(1-**p**) + (2+**n**)  (1 - 2**p**(1-**p**))  =>  
**n** = 4**p**(1-**p**) + 2 - 4**p**(1-**p**) + **n**  - 2**n** **p**(1-**p**) => сокращаем и преобразовываем  
0 = 2 - 2**n** **p**(1-**p**) => выражаем **n** через **p**  
**n** = 1/ (**p**(1-**p**))  

поэтому для **p** = 1/2, получаем n = 4. То есть если монета идеальна, то ожидаемое кол-во подбрасываний = 4 (2 эксперимента), чтобы хотя бы в одном из экспериментов получить исход 2 (орел2) или исход 3 (решка2).  

In [5]:
# функция округления в большую сторону до ближайшего четного
import math
def my_round_to_even_num(n):
    answer = math.ceil(n)
    if not answer%2: 
        return answer 
    else: 
        return answer + 1 

In [8]:

for i in range(3,11):
    p = 1 - 1/i
    n = 1/(p*(1-p))
    count_ = my_round_to_even_num(n)

    print(f'при вероятности p = {round(p, 3)}, ожидаемое кол-во подбрасываний n = {count_}')

при вероятности p = 0.667, ожидаемое кол-во подбрасываний n = 6
при вероятности p = 0.75, ожидаемое кол-во подбрасываний n = 6
при вероятности p = 0.8, ожидаемое кол-во подбрасываний n = 8
при вероятности p = 0.833, ожидаемое кол-во подбрасываний n = 8
при вероятности p = 0.857, ожидаемое кол-во подбрасываний n = 10
при вероятности p = 0.875, ожидаемое кол-во подбрасываний n = 10
при вероятности p = 0.889, ожидаемое кол-во подбрасываний n = 12
при вероятности p = 0.9, ожидаемое кол-во подбрасываний n = 12



При приближении **p** к 1 (единице), ожидаемое кол-во подбрасываний n  будет стремится к бесконечности,  
потому что мы будем получать бесконечную цепочку из "орлов" (исход 1 - "орел" - "орел") и будем вынуждены перебрасывать снова и снова)  
Но видно, что при достаточно предвзятой монете (p=0.9) ожидаемое кол-во подбрасываний 12. Это не так много с точки зрения практического применения данного способа.  

Но давайте посчитаем какова вероятность того, что мы получим череду из орлов (исход 1) или череду из решек (исход 4) и не сможем получить желаемый результат в виде исходов 2 или 3 при подбрасывании монеты **n** раз?  

При подбрасывании монеты **n** раз вероятность (**p_1_4**) получить ожидаемый исход 1 из **n** подряд идущих монет с орлом или исход 3 из **n** подряд идущих монет с решкой равен = **p**^**n** + (1-**p**)^**n**  
p_1_4 = **p**^**n** + (1-**p**)^**n**

In [39]:
for i in range(4,11, 2):
    p = 1 - 1/i
    print(f'при вероятности p = {round(p, 3)} и ')
    max = 20
    step = 4
    start = max%step
    if start == 0:
        start += step
    for n in range (start, max+1, step):
        p_1_4 = p**n + (1-p)**n
        print(f'кол-ве подбрасываний n = {n}, вероятность не получить результат = {round(p_1_4,3)}')
    print('===')

при вероятности p = 0.75 и 
кол-ве подбрасываний n = 4, вероятность не получить результат = 0.32
кол-ве подбрасываний n = 8, вероятность не получить результат = 0.1
кол-ве подбрасываний n = 12, вероятность не получить результат = 0.032
кол-ве подбрасываний n = 16, вероятность не получить результат = 0.01
кол-ве подбрасываний n = 20, вероятность не получить результат = 0.003
===
при вероятности p = 0.833 и 
кол-ве подбрасываний n = 4, вероятность не получить результат = 0.483
кол-ве подбрасываний n = 8, вероятность не получить результат = 0.233
кол-ве подбрасываний n = 12, вероятность не получить результат = 0.112
кол-ве подбрасываний n = 16, вероятность не получить результат = 0.054
кол-ве подбрасываний n = 20, вероятность не получить результат = 0.026
===
при вероятности p = 0.875 и 
кол-ве подбрасываний n = 4, вероятность не получить результат = 0.586
кол-ве подбрасываний n = 8, вероятность не получить результат = 0.344
кол-ве подбрасываний n = 12, вероятность не получить результат =

In [41]:

p = 0.99
print(f'при вероятности p = {round(p, 3)} и ')
max = 100
step = 20
start = max%step
if start == 0:
    start += step
for n in range (start, max+1, step):
    p_1_4 = p**n + (1-p)**n
    print(f'кол-ве подбрасываний n = {n}, вероятность не получить результат = {round(p_1_4,3)}')
print('===')

при вероятности p = 0.99 и 
кол-ве подбрасываний n = 20, вероятность не получить результат = 0.818
кол-ве подбрасываний n = 40, вероятность не получить результат = 0.669
кол-ве подбрасываний n = 60, вероятность не получить результат = 0.547
кол-ве подбрасываний n = 80, вероятность не получить результат = 0.448
кол-ве подбрасываний n = 100, вероятность не получить результат = 0.366
===


Это говорит нам о том, что вероятность того, что результат наступит, достаточно высока даже при высокой предвзятости монеты (p=0.99).  
Но надо согласиться, что 100 подбрасываний это уже трудоемкий процесс. Можно ли сделать этот процесс более эффективным?  

Мы можем улучшить ожидаемое количество подбрасываний монеты, если доработаем вышеописанный способ следующим образом:  
В способе который мы описали выше мы возвращаемся на пункт 1 и перебрасываем кубики, если получаем исход 1 ("орел"-"орел") или исход 4 ("решка"-"решка").  
Но мы можем дополнить вышеописанный способ:  
1. объединить событие двух исходов (исход 1 и исход 4) (ОО-РР) в исход орел3  
2. объединить событие двух исходов (исход 4 и исход 1) (РР-ОО) в исход решка3  

Из вышеописанного обоснования понятно, что вероятность исхода орел3 равна вероятности решка3 и это позволит сократить ожидаемое кол-во подбрасываний, так как мы сможем останавливать процесс быстрее не перебрасывая.  